ledger/participant-integration-api/src/main/scala/platform/store/backend/common/MeteringStorageBackendTemplate.scala

```diff
@@ -45,74 +73,111 @@ private[backend] object MeteringStorageBackendTemplate {
 
   }
 
+  /** Oracle does not understand true/false so compare against number
+    * @return 0 if the option is unset and non-zero otherwise
+    */
+  def isSet(o: Option[_]): Int = o.fold(0)(_ => 1)
+
+  /** Oracle treats zero length strings as null and then does null comparison
+    * @return 0 if the offset is beforeBegin and non-zero otherwise
+    */
+  def hasBegun(o: Offset): Int = if (o == Offset.beforeBegin) 0 else 1
+
 }
 
-private[backend] class MeteringStorageBackendReadTemplate(ledgerEndCache: LedgerEndCache)
-    extends MeteringStorageReadBackend {
+private[backend] object MeteringStorageBackendReadTemplate extends MeteringStorageReadBackend {
+
+  implicit val offsetToStatement: ToStatement[Offset] =
+    com.daml.platform.store.Conversions.OffsetToStatement
+  implicit val timestampToStatement: ToStatement[Timestamp] =
+    com.daml.platform.store.Conversions.TimestampToStatement
+  implicit val timestampParamMeta: ParameterMetaData[Timestamp] =
+    com.daml.platform.store.Conversions.TimestampParamMeta
 
-  override def transactionMetering(
+  override def reportData(
       from: Time.Timestamp,
       to: Option[Time.Timestamp],
-      applicationId: Option[ApplicationId],
-  )(connection: Connection): Vector[TransactionMetering] = {
+      maybeApplicationId: Option[ApplicationId],
+  )(connection: Connection): ReportData = {
+
+    val ledgerMeteringEnd = assertLedgerMeteringEnd(connection)
+    val participantData = participantMetering(from, to, maybeApplicationId)(connection)
+    val isFinal = to.fold(false)(ledgerMeteringEnd.timestamp >= _)
+    val data = if (isFinal) {
+      participantData
+    } else {
+      val transactionData =
+        transactionMetering(ledgerMeteringEnd.offset, to, maybeApplicationId)(connection)
+      val apps: Set[ApplicationId] = participantData.keySet ++ transactionData.keySet
+      apps.toList.map { a =>
+        a -> (participantData.getOrElse(a, 0L) + transactionData.getOrElse(a, 0L))
+      }.toMap
+    }
+
+    ReportData(data, isFinal)
+
+  }
+
+  private def transactionMetering(
+      from: Offset,
+      to: Option[Time.Timestamp],
+      appId: Option[String],
+  )(connection: Connection): Map[ApplicationId, Long] = {
+
     SQL"""
       select
         application_id,
-        action_count,
-        metering_timestamp,
-        ledger_offset
+        sum(action_count)
       from transaction_metering
-      where ledger_offset <= ${ledgerEndCache()._1.toHexString.toString}
-      and   metering_timestamp >= ${from.micros}
-      and   (${isSet(to)} = 0 or metering_timestamp < ${to.map(_.micros)})
-      and   (${isSet(applicationId)} = 0 or application_id = ${applicationId.map(_.toString)})
+      where (${hasBegun(from)} = 0 or ledger_offset > $from)
```

Comment:
`${hasBegun(from)} = 0 ` can be evaluated while constructing the sql statement.
Consider:
```
import com.daml.platform.store.backend.common.ComposableQuery.SqlStringInterpolation
val part: ComposableQuery.CompositeSql = if (!hasBegun(from)) cSQL"ledger_offset > $from" else cSQL""
...
SQL"... .$part ..."
```

Reply:
Done.

Reply:
Done.